File: leection/139.go
package leection

import "testing"

func wordBreak(s string, wordDict []string) bool {
	t := make(map[string]bool)
	for _, w := range wordDict {
		t[w] = true
	}
	dp := make([]bool, len(s)+1)
	dp[0] = true
	for i, _ := range s {
		for j := 0; j < i; j++ {
			if dp[j] && t[s[j:i-j]] {
				dp[j] = true
				break
			}
		}
	}
	return dp[len(s)]
}

func Test139(t *testing.T) {
	tt := make(map[string][]string)
	tt["leetcode"] = []string{"leet", "code"}
	for k, v := range tt {
		println(wordBreak(k, v))
	}
}

File: leection/139.cpp
#include "leection.h"
#include <unordered_set>
class Solution {
public:
    bool wordBreak(string s, vector<string>& wordDict) {
        vector<bool> dp(s.size()+1);
        unordered_set<string> sets;
        for(auto i:wordDict){
            sets.insert(i);
        }
        for(int i=0;i<s.size();i++){
            for(int j=0;j<i;j++){
                if(dp[j]&&sets.find(s.substr(j,i))!=sets.end()){
                    dp[i]=true;
                }
            }
        }
        return dp[s.size()];

    }
};
int main(){

}